core/src/test/scala/unit/kafka/server/ShareGroupHeartbeatRequestTest.scala

```diff
@@ -264,6 +264,560 @@ class ShareGroupHeartbeatRequestTest(cluster: ClusterInstance) {
+    shareGroupHeartbeatRequest = new ShareGroupHeartbeatRequest.Builder(
+      new ShareGroupHeartbeatRequestData()
+        .setGroupId("grp")
+        .setMemberId(memberId)
+        .setMemberEpoch(1), true
+    ).build()
+
+    TestUtils.waitUntilTrue(() => {
+      shareGroupHeartbeatResponse = connectAndReceive(shareGroupHeartbeatRequest)
+      shareGroupHeartbeatResponse.data.errorCode == Errors.NONE.code &&
+        shareGroupHeartbeatResponse.data.assignment == expectedAssignment
+    }, msg = s"Could not get foo partitions assigned. Last response $shareGroupHeartbeatResponse.")
+
+    // Verify the response.
+    assertEquals(2, shareGroupHeartbeatResponse.data.memberEpoch)
+
+    // Prepare the next heartbeat with a new subscribed topic.
+    shareGroupHeartbeatRequest = new ShareGroupHeartbeatRequest.Builder(
+      new ShareGroupHeartbeatRequestData()
+        .setGroupId("grp")
+        .setMemberId(memberId)
+        .setMemberEpoch(2)
+        .setSubscribedTopicNames(List("foo", "bar").asJava), true
+    ).build()
+
+    expectedAssignment = new ShareGroupHeartbeatResponseData.Assignment()
+      .setAssignedTopicPartitions(List(
+        new ShareGroupHeartbeatResponseData.TopicPartitions()
+        .setTopicId(fooId)
+        .setPartitions(List[Integer](0, 1).asJava),
+        new ShareGroupHeartbeatResponseData.TopicPartitions()
+          .setTopicId(barId)
+          .setPartitions(List[Integer](0).asJava)).asJava)
+
+    TestUtils.waitUntilTrue(() => {
+      shareGroupHeartbeatResponse = connectAndReceive(shareGroupHeartbeatRequest)
+      shareGroupHeartbeatResponse.data.errorCode == Errors.NONE.code &&
+      shareGroupHeartbeatResponse.data.assignment != null &&
+      expectedAssignment.assignedTopicPartitions.containsAll(shareGroupHeartbeatResponse.data.assignment.assignedTopicPartitions) &&
+      shareGroupHeartbeatResponse.data.assignment.assignedTopicPartitions.containsAll(expectedAssignment.assignedTopicPartitions)
+    }, msg = s"Could not get bar partitions assigned. Last response $shareGroupHeartbeatResponse.")
+
+    // Verify the response.
+    assertEquals(4, shareGroupHeartbeatResponse.data.memberEpoch)
+
+    // Prepare the next heartbeat which is empty to verify no assignment changes.
+    shareGroupHeartbeatRequest = new ShareGroupHeartbeatRequest.Builder(
+      new ShareGroupHeartbeatRequestData()
+        .setGroupId("grp")
+        .setMemberId(memberId)
+        .setMemberEpoch(4), true
+    ).build()
+
+    TestUtils.waitUntilTrue(() => {
+      shareGroupHeartbeatResponse = connectAndReceive(shareGroupHeartbeatRequest)
+      shareGroupHeartbeatResponse.data.errorCode == Errors.NONE.code
+    }, msg = s"Could not get empty heartbeat response. Last response $shareGroupHeartbeatResponse.")
+
+    // Verify the response.
+    assertEquals(4, shareGroupHeartbeatResponse.data.memberEpoch)
+
+    // Blocking the thread for 1 sec so that the heartbeat interval expires and the member needs to rejoin.
```

Comment:
This really is the session timeout I think.

Reply:
yup, I'll correct the comment, thanks